### PR TITLE
Properly delimit identifiers in DataType expression created from Type

### PR DESCRIPTION
**Pull request by losipiuk**

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/sql/analyzer/TypeSignatureTranslator.java [hunk 1]
```diff
@@ -300,6 +301,19 @@ static DataType toDataType(TypeSignature typeSignature)
         }
     }
 
+    private static boolean requiresDelimiting(String identifier)
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/TypeSignatureTranslator.java:304]: This is ok for now, but technically, every name should be delimited. Once https://github.com/trinodb/trino/issues/17 is fixed, any names coming from a connector (including column types and field names in row types) will be considered as "already canonicalized", so they'll need to be surrounded in quotes when rendering as SQL to ensure proper matching semantics.

File: core/trino-main/src/test/java/io/trino/sql/analyzer/TestTypeSignatureTranslator.java [hunk 2]
```diff
@@ -33,7 +32,7 @@ private void assertRoundTrip(String expression)
     {
         assertThat(type(expression))
                 .ignoringLocation()
-                .withComparatorForType(Comparator.comparing(identifier -> identifier.getValue().toLowerCase(Locale.ENGLISH)), Identifier.class)
+                .withComparatorForType(Comparator.comparing(Identifier::getCanonicalValue), Identifier.class)
```
Comment (martint) [inline core/trino-main/src/test/java/io/trino/sql/analyzer/TestTypeSignatureTranslator.java:35]: This changes semantics. Is it intentional? The canonical value per SQL standard rules uses upper case. For historical reasons, and until #17 is fixed, we continue to use lower case in many places.

**Review by Praveen2112 — APPROVED** [review]
After addressing @martint's [comment](https://github.com/trinodb/trino/pull/8845/files#r687286804)

File: core/trino-main/src/test/java/io/trino/sql/analyzer/TestTypeSignatureTranslator.java (hunk 2 above)
Comment (losipiuk) [reply core/trino-main/src/test/java/io/trino/sql/analyzer/TestTypeSignatureTranslator.java:35]: Does it matter here? I assumed that for sake of testing assertion whatever normalization (upper/lower) is fine. I changed to `getCanonicalValue` mostly to have different matching logic for delimited (uppercasing) and non-delimited (no uppercasing) identifiers. Without the change, the test passes for field names that require delimiting even without changes to `TypeSignatureTranslator`. I agree this is kinda a hack though. I can take that back and assume that regression testing is solely done via `testImplicitCastToRowWithFieldsRequiringDelimitation()`.

Comment (losipiuk) [reply core/trino-main/src/test/java/io/trino/sql/analyzer/TestTypeSignatureTranslator.java]: Reverted

**Comment by losipiuk** [comment]
AC @martint . Let me know if this is good to go.

**Review by martint — APPROVED** [review]

**Comment by losipiuk** [comment]
CI: https://github.com/trinodb/trino/issues/8691

**Pull request merged**